src/hdmf/spec/namespace.py

```diff
@@ -426,25 +437,71 @@ def __convert_spec_cls_keys(self, parent_cls, spec_cls, spec_dict):
-        for subspec_dict in spec_dict.get('datasets', list()):
-            self.__resolve_includes(self.__dataset_spec_cls, subspec_dict, catalog)
-
-    def __load_namespace(self, namespace, reader, resolve=True):
+    def __resolve_inc_spec(self, spec: BaseStorageSpec) -> None:
+        """Resolve the included spec into the given spec."""
+        # get the spec for the included type regardless of namespace
+        included_spec = self.get_spec_for_type(spec.data_type_inc)
+
+        # resolve the included spec only after the included spec has been resolved
+        if included_spec.resolved:
+            spec.resolve_inc_spec(included_spec)
+
+    def __resolve_spec(self, spec: BaseStorageSpec) -> None:
+        """Resolve the given spec by resolving its included spec and all its subspecs."""
+        # NOTE: resolution needs to happen at the namespace catalog level because the included spec
+        # may be in a different namespace
+        if not spec.resolved:
+            # resolve the included spec before the subspecs because the subspecs may refine fields inherited from the
+            # included spec
+            if spec.data_type_inc is not None and not spec.inc_spec_resolved:
+                self.__resolve_inc_spec(spec)
+
+            if spec.data_type_inc is None or spec.inc_spec_resolved:
+                if isinstance(spec, DatasetSpec):
+                    # dataset spec has no subspecs so if it has been resolved with the included spec, it is done
+                    spec.resolved = True
+                    return
+
+                # group spec - try to resolve all subspecs and note if they are all resolved
+                all_subspecs_resolved = True
+                for subspec in (spec.groups + spec.datasets):
+                    self.__resolve_spec(subspec)
+                    if not subspec.resolved:
+                        all_subspecs_resolved = False
+
+                if all_subspecs_resolved:
+                    spec.resolved = True
+
+    def resolve_all_specs(self) -> None:
+        """Resolve all specs in the catalog."""
+        # get a set of all unresolved data type specs
+        unresolved_data_type_specs = set()
+        for namespace in self.__namespaces.values():
+            for type_name in namespace.catalog.get_registered_types():
+                spec = namespace.catalog.get_spec(type_name)
+                if not spec.resolved:
+                    unresolved_data_type_specs.add(spec)
+
+        # iteratively resolve specs until all are resolved or we reach max passes
+        num_passes = 0
+        max_passes = 10
+        while unresolved_data_type_specs and num_passes < max_passes:
+            num_passes += 1
+            specs_to_process = copy(unresolved_data_type_specs)
+            for spec in specs_to_process:
+                self.__resolve_spec(spec)
+                if spec.resolved:
+                    unresolved_data_type_specs.remove(spec)
```

Comment:
Comments from discussion with @stephprince. The graph might be a little clearer to understand and more efficient. This involves more code though.